pkg/cmd/cli/describe/projectstatus.go

```diff
@@ -581,3 +569,94 @@ func describeServicePorts(spec kapi.ServiceSpec) string {
 		return " ports " + strings.Join(pairs, ", ")
 	}
 }
+
+func loadServices(g osgraph.Graph, graphLock sync.Mutex, namespace string, kclient kclient.Interface, client client.Interface) error {
+	svcs, err := kclient.Services(namespace).List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	graphLock.Lock()
+	defer graphLock.Unlock()
```

Comment:
By default - ALWAYS defer your unlock.  Only in extremely rare cases should
the code be structured so that unlock is not a defer.

If the code panics and you haven't unlocked, your process is broken (likely
permanently, in unexpected ways).  It's just so much more important to
ensure locks are paired

On Jul 12, 2015, at 8:39 AM, Jordan Liggitt <notifications@github.com>
wrote:

In pkg/cmd/cli/describe/projectstatus.go
<https://github.com/openshift/origin/pull/3621#discussion_r34421302>:

> @@ -581,3 +569,94 @@ func describeServicePorts(spec kapi.ServiceSpec) string {
>  		return " ports " + strings.Join(pairs, ", ")
>  	}
>  }
> +
> +func loadServices(g osgraph.Graph, graphLock sync.Mutex, namespace string, kclient kclient.Interface, client client.Interface) error {
> +	svcs, err := kclient.Services(namespace).List(labels.Everything())
> +	if err != nil {
> +		return err
> +	}
> +
> +	graphLock.Lock()
> +	defer graphLock.Unlock()

They're not expensive relative to the network cal we're about to make.
Deferring immediately after locking gives good locality of reference,
ensures the unlock is called even in case of a panic, etc

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3621/files#r34421302>.